direct/functionals/hfen.py

```diff
@@ -0,0 +1,320 @@
+from typing import Optional
+
+import numpy as np
+import torch
+from torch import nn
+
+__all__ = ["hfen_l1", "hfen_l2", "HFENLoss", "HFENL1Loss", "HFENL2Loss"]
+
+
+def get_log_kernel2d(kernel_size: int | list[int] = 5, sigma: Optional[float | list[float]] = None) -> torch.Tensor:
+    """Generates a 2D LoG (Laplacian of Gaussian) kernel.
+
+    Parameters
+    ----------
+    kernel_size : int or list of ints
+        Size of the kernel. Default: 5.
+    sigma : float or list of floats
+        Standard deviation(s) for the Gaussian distribution. Default: None.
+
+    Returns
+    -------
+    torch.Tensor: Generated LoG kernel.
+    """
+    dim = 2
+    if not kernel_size and sigma:
+        kernel_size = np.ceil(sigma * 6)
+        kernel_size = [kernel_size] * dim
+    elif kernel_size and not sigma:
+        sigma = kernel_size / 6.0
+        sigma = [sigma] * dim
+
+    if isinstance(kernel_size, int):
+        kernel_size = [kernel_size - 1] * dim
+    if isinstance(sigma, float):
+        sigma = [sigma] * dim
+
+    grids = torch.meshgrid([torch.arange(-size // 2, size // 2 + 1, 1) for size in kernel_size], indexing="ij")
+
+    kernel = 1
+    for size, std, mgrid in zip(kernel_size, sigma, grids):
+        kernel *= torch.exp(-(mgrid**2 / (2.0 * std**2)))
+
+    final_kernel = (
+        kernel
+        * ((grids[0] ** 2 + grids[1] ** 2) - (2 * sigma[0] * sigma[1]))
+        * (1 / ((2 * math.pi) * (sigma[0] ** 2) * (sigma[1] ** 2)))
+    )
+
+    final_kernel = -final_kernel / torch.sum(final_kernel)
+
+    return final_kernel
+
+
+def compute_padding(kernel_size: int | list[int] = 5) -> int | tuple[int, ...]:
+    """Computes padding tuple based on the kernel size.
+
+    For square kernels, pad can be an int, else, a tuple with an element for each dimension.
+
+    Parameters
+    ----------
+    kernel_size : int or list of ints
+        Size(s) of the kernel.
+
+    Returns
+    -------
+    int or tuple of ints
+        Computed padding.
+    """
+    if isinstance(kernel_size, int):
+        return kernel_size // 2
+    # Else assumed a list
+    computed = [k // 2 for k in kernel_size]
+    out_padding = []
+    for i, _ in enumerate(kernel_size):
+        computed_tmp = computed[-(i + 1)]
+        if kernel_size[i] % 2 == 0:
+            padding = computed_tmp - 1
+        else:
+            padding = computed_tmp
+        out_padding.append(padding)
+        out_padding.append(computed_tmp)
+    return tuple(out_padding)
+
+
+class HFENLoss(nn.Module):
+    """High Frequency Error Norm (HFEN) Loss as defined in _[1].
+
+    Calculates:
+
+    .. math::
+
+        || \text{LoG}(x_\text{rec}) - \text{LoG}(x_\text{tar}) ||_C
+
+    Where C can be any norm, LoG is the Laplacian of Gaussian filter, and :math:`x_\text{rec}), \text{LoG}(x_\text{tar}`
+    are the reconstructed inp and target images.
+    If normalized it scales it by :math:`|| \text{LoG}(x_\text{tar}) ||_C`.
+
+    Code was borrowed and adapted from _[2].
+
```

Comment:
Mention the license of where you derived it, and if it's compatible with the current one.

Reply:
There is no license present. I will just mention that (no license exists)